### PR TITLE
Gas metering

### PR DESCRIPTION
**Pull request by colmazia**
Fixed: update gas metering

## Implementation details
- update owasm gas metering to use wasmer middleware
- remove old gas metering
- update wasmer to 2.3.0
- **comment out vrf as gmp lib has some issues

## Please ensure the following requirements are met before submitting a pull request:

- [x] The pull request is targeted against the correct target branch
- [x] The pull request is linked to an issue with appropriate discussion and an accepted design OR is linked to a spec that describes the work.
- [x] The pull request includes a description of the implementation/work done in detail.
- [x] The pull request includes any and all appropriate unit/integration tests
- [ ] You have added a relevant changelog entry to `CHANGELOG_UNRELEASED.md`
- [x] You have re-reviewed the files affected by the pull request (e.g. using the `Files changed` tab in the Github PR explorer)

### CODE REVIEW

File: packages/vm/src/lib.rs [hunk 1]
```diff
@@ -73,11 +79,11 @@ fn inject_stack_height(module: Module) -> Result<Module, Error> {
         .map_err(|_| Error::StackHeightInjectionError)
 }
 
-fn inject_gas(module: Module) -> Result<Module, Error> {
-    // Simple gas rule. Every opcode and memory growth costs 1 gas.
-    let gas_rules = rules::Set::new(1, Default::default()).with_grow_cost(1);
-    pwasm_utils::inject_gas_counter(module, &gas_rules).map_err(|_| Error::GasCounterInjectionError)
-}
+// fn inject_gas(module: Module) -> Result<Module, Error> {
```
Comment (RogerKSI) [inline packages/vm/src/lib.rs]: please remove it. think we don't need it anymore.

File: packages/vm/src/vm.rs [hunk 2]
```diff
@@ -122,6 +109,56 @@ where
...
+
+    pub fn set_gas_left(&self, new_value: u64) {
+        self.with_wasmer_instance(|instance| {
+            set_remaining_points(instance, new_value);
+            Ok(())
+        })
+        .expect("Wasmer instance is not set. This is a bug in the lifecycle.")
+    }
+
+    pub fn decrease_gas_left(&self, gas_used: u32) -> Result<(), Error> {
```
Comment (RogerKSI) [inline packages/vm/src/vm.rs]: Currently, we use u64 for gas. Can we use u64 for gas_used?

File: packages/vm/src/lib.rs [hunk 3]
```diff
@@ -135,33 +140,38 @@ fn require_mem_range(max_range: usize, require_range: usize) -> Result<(), Error
...
+    let engine = Universal::new(compiler).engine();
+    let store = Store::new(&engine);
 
     let import_object = imports! {
         "env" => {
-            "gas" => Function::new_native_with_env(&store, owasm_env.clone(), |env: &Environment<E>, gas: u32| {
-                env.with_mut_vm(|vm| {
-                    vm.consume_gas(gas)
-                })
+            "gas" => Function::new_native_with_env(&store, owasm_env.clone(), |_env: &Environment<E>, _gas: u32| {
```
Comment (RogerKSI) [inline packages/vm/src/lib.rs]: please add env.decrease_gas_left(1) here

**Review by RogerKSI — CHANGES_REQUESTED** [review]

**Review by RogerKSI — APPROVED** [review]

File: packages/vm/src/lib.rs [hunk 4]
```diff
@@ -9,12 +9,18 @@ pub use error::Error;
...
+use pwasm_utils::{self};
 
-use wasmer::{imports, wasmparser, Function, Singlepass, Store, JIT};
+use wasmer::Singlepass;
 
-use owasm_crypto::{ecvrf};
+use wasmer::{imports, wasmparser, wasmparser::Operator, CompilerConfig, Function, Store};
+use wasmer_middlewares::Metering;
+
+// use owasm_crypto::ecvrf;
```
Comment (RogerKSI) [inline packages/vm/src/lib.rs:23]: please remove "env.ecvrf_verify" in SUPPORTED_IMPORTS.

**Review by RogerKSI — CHANGES_REQUESTED** [review]

File: packages/vm/src/vm.rs [hunk 5]
```diff
@@ -122,6 +109,56 @@ where
         data.wasmer_instance = instance;
     }
 
+    pub fn with_wasmer_instance<C, R>(&self, callback: C) -> VmResult<R>
```
Comment (RogerKSI) [inline packages/vm/src/vm.rs:112]: please use our own result and error.

File: packages/vm/Cargo.toml [hunk 6]
```diff
@@ -13,8 +13,9 @@ assert_matches = "1.3.0"
...
 parity-wasm = "0.41"
 pwasm-utils = "0.12"
-wasmer = { version = "1.0.2", default-features = false, features = ["jit", "singlepass", "compiler"] }
-owasm-crypto = { path = "../crypto", version = "0.1.13" }
+wasmer = { version = "2.3.0", default-features = false, features = ["singlepass", "compiler", "universal"] }
+wasmer-middlewares = "2.3.0"
+# owasm-crypto = { path = "../crypto", version = "0.1.13" }
 tempfile = "3.1.0"
 clru = "0.2.0"
 cosmwasm-vm = "0.13.2"
```
Comment (RogerKSI) [inline packages/vm/Cargo.toml:21]: Can we upgrade the version?

**Review by RogerKSI — CHANGES_REQUESTED** [review]

File: packages/vm/src/vm.rs (hunk 5 above)
Comment (colmazia) [reply packages/vm/src/vm.rs:112]: Will be fixed in refactor

File: packages/vm/Cargo.toml (hunk 6 above)
Comment (colmazia) [reply packages/vm/Cargo.toml:21]: Will be fixed in refactor

File: packages/vm/src/vm.rs [hunk 7]
```diff
@@ -122,6 +109,56 @@ where
...
+
+    pub fn set_gas_left(&self, new_value: u64) {
+        self.with_wasmer_instance(|instance| {
+            set_remaining_points(instance, new_value);
+            Ok(())
+        })
+        .expect("Wasmer instance is not set. This is a bug in the lifecycle.")
+    }
+
+    pub fn decrease_gas_left(&self, gas_used: u64) -> Result<(), Error> {
```
Comment (taobun) [inline packages/vm/src/vm.rs]: rename `gas_used` to `gas`

**Review by RogerKSI — APPROVED** [review]

**Review by taobun — APPROVED** [review]

**Pull request merged**